File: sdk/javascript/src/exceptions.js
/**
 * OpenClaw 量化交易平台 JavaScript SDK - 自定义异常
 */

class OpenClawError extends Error {
  /**
   * @param {string} message - 错误消息
   * @param {number} [code] - 错误码
   * @param {object} [details] - 错误详情
   */
  constructor(message, code = null, details = {}) {
    super(message)
    this.name = 'OpenClawError'
    this.code = code
    this.details = details
  }

  toString() {
    return this.code ? `[${this.code}] ${this.message}` : this.message
  }
}

class AuthenticationError extends OpenClawError {
  constructor(message = '认证失败，请检查 API Key 和 Secret', details) {
    super(message, 401, details)
    this.name = 'AuthenticationError'
  }
}

class RateLimitError extends OpenClawError {
  /**
   * @param {string} [message] - 错误消息
   * @param {number} [retryAfter] - 重试等待秒数
   */
  constructor(message = '请求频率超限，请稍后重试', retryAfter = null) {
    super(message, 429, { retryAfter })
    this.name = 'RateLimitError'
    this.retryAfter = retryAfter
  }
}

class APIError extends OpenClawError {
  /**
   * @param {string} [message] - 错误消息
   * @param {number} [statusCode] - HTTP 状态码
   */
  constructor(message = 'API 请求错误', statusCode = null) {
    super(message, statusCode, {})
    this.name = 'APIError'
    this.statusCode = statusCode
  }
}

class ConnectionError extends OpenClawError {
  constructor(message = '无法连接到 OpenClaw 服务') {
    super(message, 0, {})
    this.name = 'ConnectionError'
  }
}

module.exports = {
  OpenClawError,
  AuthenticationError,
  RateLimitError,
  APIError,
  ConnectionError,
}

File: sdk/python/openclaw_sdk/exceptions.py
"""
OpenClaw 量化交易平台 Python SDK - 自定义异常
"""


class OpenClawError(Exception):
    """OpenClaw SDK 基础异常"""

    def __init__(self, message: str, code: int = None, details: dict = None):
        self.message = message
        self.code = code
        self.details = details or {}
        super().__init__(self.message)

    def __str__(self):
        result = self.message
        if self.code:
            result = f"[{self.code}] {result}"
        return result


class AuthenticationError(OpenClawError):
    """认证失败异常"""

    def __init__(self, message: str = "认证失败，请检查 API Key 和 Secret", **kwargs):
        super().__init__(message=message, code=401, **kwargs)


class RateLimitError(OpenClawError):
    """限流异常"""

    def __init__(self, message: str = "请求频率超限，请稍后重试", retry_after: int = None, **kwargs):
        self.retry_after = retry_after
        super().__init__(message=message, code=429, **kwargs)


class APIError(OpenClawError):
    """API 错误异常"""

    def __init__(self, message: str = "API 请求错误", status_code: int = None, **kwargs):
        self.status_code = status_code
        super().__init__(message=message, code=status_code, **kwargs)


class ConnectionError(OpenClawError):
    """连接错误异常"""

    def __init__(self, message: str = "无法连接到 OpenClaw 服务", **kwargs):
        super().__init__(message=message, code=0, **kwargs)
